File: python/char_count.py


from collections import Counter
from string import ascii_letters


def char_count(string):
    """
    This function returns a list of pairs of characters and their counts sorted lexically.

    Args:
        string (str): A string of characters.

    Returns:
        list[str]: A list of pairs of characters and their counts sorted lexically.
    """

    valid_chars = set(ascii_letters)

    chars = [char.lower() for char in string if char in valid_chars]
    counts = sorted(Counter(chars).items(), key=lambda x: x[1], reverse=True)

    return [[k, v] for k, v in counts]

File: js/charCount.js
/**
 * This function returns an array of characters in a word and their counts.
 *
 * @param {String} word The word to return counts of
 * @return {Object} An object with the counts of the word
 */
const charCount = (word) => {
  const validChars = new Set(
    Array.from("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
  );
  const chars = Array.from(word)
    .filter((char) => validChars.has(char))
    .map((char) => char.toLowerCase());

  const counts = {};
  for (const char of chars) {
    counts[char] = counts[char] || 0;
    counts[char]++;
  }

  return Object.keys(counts)
    .sort((a, b) => counts[b] - counts[a])
    .map((char) => [char, counts[char]]);
};

module.exports = { charCount };
